tests/test_ui_web.py

```
@@ -251,7 +278,131 @@ async def test_get_ui_html_filesystem_cache_hit(monkeypatch: pytest.MonkeyPatch,
+    """The temp file handle is closed before `os.replace`, so the rename can't fail on Windows.
+
+    Windows refuses to replace a file that still has an open handle, so an `os.replace` fired while
+    the `NamedTemporaryFile` handle is open would break the atomic write there. Capturing the temp
+    file object and asserting it is already closed when `os.replace` runs pins the close-before-replace
+    ordering on every platform (POSIX allows renaming an open file, so it would otherwise hide the bug).
+    """
+    cache_file = tmp_path / f'{app_module.CHAT_UI_VERSION}.html'
+    temp_files: list[IO[bytes]] = []
+    real_named_temporary_file = app_module.tempfile.NamedTemporaryFile
+
+    def capturing_named_temporary_file(*, dir: Path, prefix: str, delete: bool) -> IO[bytes]:
```

Ruff (0.15.20)

[error] Function argument `dir` is shadowing a Python builtin

(A002)